### ISSUE TITLE
Re-running H5Coro.readDatasets returns UnboundLocalError 

### ISSUE THREAD

**Issue by rwegener2** [ICESat2-SlideRule/h5coro#10]
After the initial setup:
```python
import earthaccess
from h5coro import h5coro, s3driver

s3url_atl03 = 'nsidc-cumulus-prod-protected/ATLAS/ATL03/006/2019/11/30/ATL03_20191130112041_09860505_006_01.h5'

# authenticate
auth = earthaccess.login()
s3_creds = auth.get_s3_credentials(daac='NSIDC')

h5obj = h5coro.H5Coro(s3url_atl03, 
                      s3driver.S3Driver,
                      credentials={"aws_access_key_id": s3_creds["accessKeyId"],
                                 "aws_secret_access_key": s3_creds["secretAccessKey"],
                                 "aws_session_token": s3_creds["sessionToken"], })

out = h5obj.readDatasets(datasets=['gt1l/heights/ph_id_pulse'])
print(out)
```
The first time running `out = h5obj.readDatasets(datasets=['gt1l/heights/ph_id_pulse'])` goes as expected. Re-running the that same line returns:
```python
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
Cell In[62], line 1
----> 1 out = h5obj.readDatasets(datasets=['gt1l/heights/ph_id_pulse'])

File ~/h5coro/h5coro/h5coro.py:127, in H5Coro.readDatasets(self, datasets, block, earlyExit, metaOnly, enableAttributes)
    124         datasetTable[dataset["dataset"]] = dataset
    126 # return promise
--> 127 return H5Promise(self, datasetTable, block, earlyExit=earlyExit, metaOnly=metaOnly, enableAttributes=enableAttributes)

File ~/h5coro/h5coro/h5promise.py:83, in H5Promise.__init__(self, resourceObject, datasetTable, block, earlyExit, metaOnly, enableAttributes)
     81 # wait for datasets to be populated OR populate datasets in the background
     82 if block:
---> 83     resultThread(self, futures)
     84 else:
     85     Thread(target=resultThread, args=(self,futures), daemon=True).start()

File ~/h5coro/h5coro/h5promise.py:54, in resultThread(promise, futures)
     52 def resultThread(promise, futures):
     53     for future in as_completed(futures):
---> 54         h5dataset = future.result()
     55         promise.conditions[h5dataset.dataset].acquire()
     56         promise.datasets[h5dataset.dataset] = h5dataset

File ~/envs/h5coro-dev/lib/python3.11/concurrent/futures/_base.py:449, in Future.result(self, timeout)
    447     raise CancelledError()
    448 elif self._state == FINISHED:
--> 449     return self.__get_result()
    451 self._condition.wait(timeout)
    453 if self._state in [CANCELLED, CANCELLED_AND_NOTIFIED]:

File ~/envs/h5coro-dev/lib/python3.11/concurrent/futures/_base.py:401, in Future.__get_result(self)
    399 if self._exception:
    400     try:
--> 401         raise self._exception
    402     finally:
    403         # Break a reference cycle with the exception in self._exception
    404         self = None

File ~/envs/h5coro-dev/lib/python3.11/concurrent/futures/thread.py:58, in _WorkItem.run(self)
     55     return
     57 try:
---> 58     result = self.fn(*self.args, **self.kwargs)
     59 except BaseException as exc:
     60     self.future.set_exception(exc)

File ~/h5coro/h5coro/h5promise.py:47, in datasetThread(resourceObject, dataset, startRow, numRows, earlyExit, metaOnly, enableAttributes)
     45 def datasetThread(resourceObject, dataset, startRow=0, numRows=H5Dataset.ALL_ROWS, *, earlyExit, metaOnly, enableAttributes):
     46     try:
---> 47         return H5Dataset(resourceObject, dataset, startRow, numRows, earlyExit=earlyExit, metaOnly=metaOnly, enableAttributes=enableAttributes)
     48     except RuntimeError as e:
     49         logger.warning(f'H5Coro encountered error reading {dataset}: {e}')

File ~/h5coro/h5coro/h5dataset.py:220, in H5Dataset.__init__(self, resourceObject, dataset, startRow, numRows, makeNull, earlyExit, metaOnly, enableAttributes)
    218 # read b-tree
    219 self.pos = self.meta.address
--> 220 self.readBTreeV1(buffer, buffer_offset, dataset_level)
    222 # check need to flatten chunks
    223 flatten = False

UnboundLocalError: cannot access local variable 'dataset_level' where it is not associated with a value
```
Is this expected? Totally fine if this is expected functionality, just checking in to make sure I'm not doing something wrong.

**Comment by jpswinski** [comment]
I was able to recreate the error and it was a bug in the h5dataset code.  I pushed a fix: [4878f05](https://github.com/ICESat2-SlideRule/h5coro/commit/4878f05023e59cd117a3ba12392ce83cec42d65e)

**Comment by rwegener2** [comment]
This fixed it! Thanks @jpswinski! 🎉
